test/leafnode_test.go

```diff
@@ -564,6 +568,65 @@ func TestLeafNodeNoEcho(t *testing.T) {
 	expectNothing(t, lc)
 }
 
+func TestLeafNodeHeaderSupport(t *testing.T) {
+	srvA, optsA := runLeafServer()
+	defer srvA.Shutdown()
+
+	srvB, optsB := runSolicitLeafServer(optsA)
+	defer srvB.Shutdown()
+
+	clientA := createClientConn(t, optsA.Host, optsA.Port)
+	defer clientA.Close()
+
+	clientB := createClientConn(t, optsB.Host, optsB.Port)
+	defer clientB.Close()
+
+	sendA, expectA := setupHeaderConn(t, clientA)
+	sendA("SUB foo bar 22\r\n")
+	sendA("SUB bar 11\r\n")
+	sendA("PING\r\n")
+	expectA(pongRe)
+
+	if err := checkExpectedSubs(3, srvB); err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	sendB, expectB := setupHeaderConn(t, clientB)
+	// Can not have \r\n in payload fyi for regex.
+	// With reply
+	sendB("HPUB foo reply 12 14\r\nK1:V1,K2:V2 ok\r\n")
+	sendB("PING\r\n")
+	expectB(pongRe)
+
+	expectHeaderMsgs := expectHeaderMsgsCommand(t, expectA)
+	matches := expectHeaderMsgs(1)
+	checkHmsg(t, matches[0], "foo", "22", "reply", "12", "14", "K1:V1,K2:V2 ", "ok")
+
+	// Without reply
+	sendB("HPUB foo 12 14\r\nK1:V1,K2:V2 ok\r\n")
+	sendB("PING\r\n")
+	expectB(pongRe)
+
+	matches = expectHeaderMsgs(1)
+	checkHmsg(t, matches[0], "foo", "22", "", "12", "14", "K1:V1,K2:V2 ", "ok")
+
+	// Without queues or reply
```

Comment:
same test as above. should this be without reply but queue and have a |?

Reply:
and some test doing both?

Reply:
I have two subs, one is "foo" with a queue and "bar" does not.

Reply:
So when I send to foo testing queues, and when I send to bar am not. Easy to miss though.. ;)

Reply:
shouldn't I see tests with sendB("HPUB that contain + and |?

Reply:
as per line 1493 in leafnode.go?

Reply:
Parsing HMSG protocol is reused from routes tests and parser tests. This test does test that the leafnode can receive HMSG protos though.